ArmyOfOnes/ArmyOfOnes/ViewControllers/CurrenciesListViewController.swift

```diff
@@ -0,0 +1,216 @@
+        tableView.delegate = self
+        tableView.dataSource = self
+        
+        title = listTitle
+        tableView?.register(UITableViewCell.self, forCellReuseIdentifier: cellIdentifier)
+        
+        // Gesture that recognices tap to hide keyboard
+        let tapGesture = UITapGestureRecognizer(target: self, action:#selector(CurrenciesListViewController.endEditing(_:)))
+        tapGesture.cancelsTouchesInView = false
+        tableView.addGestureRecognizer(tapGesture);
+        
+        setQuantityInputLayout()
+        setupRates()
+        
+        view.setNeedsUpdateConstraints()
+    }
+    
+    func setQuantityInputLayout() -> Void {
+        
+        quantityInput.delegate = self
+        quantityInput.addTarget(self, action: #selector(CurrenciesListViewController.textFieldChanged(_:)), for: UIControlEvents.editingChanged)
+        navigationController?.view.insertSubview(quantityInput, belowSubview: navigationController!.navigationBar)
+        
+        //sets format number style to currency
+        currencyFormatter.numberStyle = NumberFormatter.Style.currency
+    }
+    
+    func showNoConnectionMessage() -> Void {
+        //shows an alert message when there is no internet connection
+        let alert = UIAlertController(title: "Connection error", message: "Make sure you are connected to the Internet", preferredStyle: UIAlertControllerStyle.alert)
+        alert.addAction(UIAlertAction(title: "OK", style: UIAlertActionStyle.default, handler: nil))
+        present(alert, animated: true, completion: nil)
+    }
+    
+    func setupRates() -> Void {
+        
+        let currencyController = CurrencyController()
+        currencyController.fetchRates(Country.allCountries) {(fetchedRates, error) -> Void in
+            
+            
+            guard let _ = error else  {
+                
+                guard let ratesList = fetchedRates else {
+                    print("rates list is not defined")
+                    return
+                }
+                
+                self.rates = ratesList
+                self.reloadRates()
+                
+                return
+            }
+            
+            self.showNoConnectionMessage()
+            
+        }
+
+    }
+    
+    func reloadRates() -> Void {
+        currencyRatesList.removeAll()
+        currencyCountries.removeAll()
+        
+        for rate in rates {
+            let countryValue = rate.country.rawValue
+            let rateValue = rate.value * dollarQuantity
+            currencyFormatter.currencyCode = countryValue
+            let convertedRateValue = NSNumber(value: rateValue)
+
+            if let currencyFormatedValue = currencyFormatter.string(from: convertedRateValue) {
+                currencyRatesList.append("\(countryValue) \(currencyFormatedValue)")
+                currencyCountries.append("\(countryValue)")
+            }
+            DispatchQueue.main.async{
```

Comment:
Why are you explicitly running this on the main thread? I'd assume the whole method should already be running on the main thread.